src/os/bluestore/bluestore_types.cc

```diff
@@ -335,6 +335,241 @@ ostream& operator<<(ostream& out, const bluestore_extent_ref_map_t& m)
+      offset += (phase ? au_size - phase : au_size);
+    }
+  }
+}
+
+bool bluestore_blob_use_tracker_t::put(
+  uint32_t offset, uint32_t length,
+  std::function<void(uint32_t*, uint32_t*)> fallback_cb,
+  set<uint32_t> *release_units)
+{
+  if (release_units) {
+    release_units->clear();
+  }
+  if (!num_au) {
+    if (total_bytes == length) {
+      total_bytes = 0;
+      return true;
+    }
+    uint32_t _au_size = 0, _num_au = 0;
+    fallback_cb(&_num_au, &_au_size);
+    if (_num_au && _au_size) { // we might be forbidden to fallback to partial
+      fall_back_to_partial(_num_au, _au_size);
+    } else {
+      total_bytes -= length;
+    }
+  }
+  bool maybe_empty = true;
+  if (num_au) {
+    assert(au_size != 0);
+    auto end = offset + length;
+    while (offset < end) {
+      auto phase = offset % au_size;
+      size_t pos = offset / au_size;
+      auto diff = MIN(au_size - phase, end - offset);
+      assert(diff <= bytes_per_au[pos]);
+      bytes_per_au[pos] -= diff;
+      offset += (phase ? au_size - phase : au_size);
+      if (bytes_per_au[pos] == 0) {
+	if (release_units) {
+	  release_units->insert(pos * au_size);
+	}
+      } else {
+	maybe_empty = false; // micro optimization detecting we aren't empty 
+	                     // even in the affected extent
+      }
+    }
+  }
+  bool empty = maybe_empty ? !is_not_empty() : false;
+  if (empty && release_units) {
+    release_units->clear();
+  }
+  return empty;
+}
+
+void bluestore_blob_use_tracker_t::split(
+  uint32_t blob_offset,
+  bluestore_blob_use_tracker_t* r)
+{
+  assert(r->bytes_per_au == 0); // target is empty
+
+  if (!num_au) {
+    if (blob_offset < total_bytes) {
+      r->total_bytes = total_bytes - blob_offset;
+      total_bytes = blob_offset;
+    } else {
+      r->total_bytes = 0;
+    }
+  } else {
+    assert(au_size);
+    assert((blob_offset % au_size) == 0 );
+    assert(blob_offset <= num_au * au_size);
+    
+    uint32_t new_num_au = blob_offset / au_size;
+
+    r->fall_back_to_partial(num_au - new_num_au, au_size);
+    for (auto i = new_num_au; i < num_au; ++i) {
+      r->bytes_per_au[i - new_num_au] = bytes_per_au[i];
+      bytes_per_au[i] = 0;
+    }
+    if (new_num_au == num_au) {
+      clear();
+    } else {
+      num_au = new_num_au;
+    }
+  }
+}
+
+bool bluestore_blob_use_tracker_t::equal(
+  const bluestore_blob_use_tracker_t& other) const
+{
+  if (!num_au && !other.num_au) {
+    return total_bytes == other.total_bytes;
+  } else if (num_au && other.num_au) {
+    auto cnt = MIN(num_au, other.num_au);
+    for (size_t i = 0; i < cnt; i++) {
+      if (bytes_per_au[i] != other.bytes_per_au[i]) {
+        return false;
+      }
+    }
+    for (size_t i = cnt; i < MAX(num_au, other.num_au); i++) {
+      if (i < num_au && bytes_per_au[i]) {
+        return false;
+      }
+      if (i < other.num_au && other.bytes_per_au[i]) {
+        return false;
+      }
+    }
+    return true;
```

Comment:
I don't understand this block.  Won't we return false if au_size doesn't match?  (The first bytes_per_au values won't be the same.)  And what would it mean to compare equality when num_au doesn't match?

It looks like this is only used from fsck, in which case the size and au_size and num_au will match, right?